forge/src/executor/mod.rs

```diff
@@ -0,0 +1,353 @@
+    /// The status of the call
+    pub status: Return,
+    /// The decoded result of the call
+    pub result: D,
+    /// The gas used for the call
+    pub gas: u64,
+    /// The logs emitted during the call
+    pub logs: Vec<RawLog>,
+    /// The changeset of the state.
+    ///
+    /// This is only present if the changed state was not committed to the database (i.e. if you
+    /// used `call` and `call_raw` not `call_committing` or `call_raw_committing`).
+    pub state_changeset: Option<HashMap<Address, Account>>,
+}
+
+/// The result of a raw call.
+#[derive(Debug)]
+pub struct RawCallResult {
+    /// The status of the call
+    pub status: Return,
+    /// The raw result of the call
+    pub result: Bytes,
+    /// The gas used for the call
+    pub gas: u64,
+    /// The logs emitted during the call
+    pub logs: Vec<RawLog>,
+    /// The changeset of the state.
+    ///
+    /// This is only present if the changed state was not committed to the database (i.e. if you
+    /// used `call` and `call_raw` not `call_committing` or `call_raw_committing`).
+    pub state_changeset: Option<HashMap<Address, Account>>,
+}
+
+pub struct Executor<DB: DatabaseRef> {
+    // Note: We do not store an EVM here, since we are really
+    // only interested in the database. REVM's `EVM` is a thin
+    // wrapper around spawning a new EVM on every call anyway,
+    // so the performance difference should be negligible.
+    //
+    // Also, if we stored the VM here we would still need to
+    // take `&mut self` when we are not committing to the database, since
+    // we need to set `evm.env`.
+    db: CacheDB<DB>,
+    env: Env,
+    // TODO: Here we are going to store information about the enabled inspectors, or just the
+    // meta-inspector.
+    // NOTE: It is important that the inspector gets a new state every time.
+    //inspector: LogCollector,
+}
+
+impl<DB> Executor<DB>
+where
+    DB: DatabaseRef,
+{
+    pub fn new(inner_db: DB, env: Env) -> Self {
+        Executor { db: CacheDB::new(inner_db), env }
+    }
+
+    /// Set the balance of an account.
+    pub fn set_balance(&mut self, address: Address, amount: U256) {
+        let mut account = self.db.basic(address);
+        account.balance = amount;
+
+        self.db.insert_cache(address, account);
+    }
+
+    /// Calls the `setUp()` function on a contract.
+    pub fn setup(
+        &mut self,
+        address: Address,
+    ) -> std::result::Result<(Return, Vec<RawLog>), EvmError> {
+        let CallResult { status, logs, .. } = self.call_committing::<(), _, _>(
+            Address::zero(),
+            address,
+            "setUp()",
+            (),
+            0.into(),
+            None,
+        )?;
+        Ok((status, logs))
+    }
+
+    /// Performs a call to an account on the current state of the VM.
+    ///
+    /// The state after the call is persisted.
+    pub fn call_committing<D: Detokenize, T: Tokenize, F: IntoFunction>(
```

Comment:
There are two variants of calls, comitting and non-comitting. Comitting ones are pretty much only used during test setup. The non-comitting ones return a state changeset. I've been thinking I might be able to dedupe it somewhat since they do almost the same thing (I think we can call `db.commit(changeset)` to do what `transact_commit` essentially does)